### PR TITLE
code-server: 3.4.1 -> 3.6.0

### PR DESCRIPTION
**Pull request by dguenther**
###### Motivation for this change

Updates code-server to a newer version. It now bundles a Go executable for connecting to Coder Cloud. Also cleaned up a few instances where yarn and npm packages were attempting to hit the network, causing the sandboxed build to fail.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/code-server/default.nix [hunk 1]
```diff
@@ -12,17 +12,31 @@ let
 
 in stdenv.mkDerivation rec {
   pname = "code-server";
-  version = "3.4.1";
-  commit = "d3773c11f147bdd7a4f5acfefdee23c26f069e76";
+  version = "3.6.0";
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server/default.nix:15]: and line 21: The version number probably shouldn't be define in two places. I think we can safely move it in the let.

**Review by SuperSandro2000 — COMMENTED** [review]
Result of `nixpkgs-review pr 101628` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>

  - code-server
</details>

File: pkgs/servers/code-server/default.nix (hunk 1 above)
Comment (dguenther) [reply pkgs/servers/code-server/default.nix:15]: Thanks for the catch! Updated to reference the derivation's version, let me know if you had something else in mind

File: pkgs/servers/code-server/default.nix [hunk 2]
```diff
@@ -12,17 +12,31 @@ let
...
   };
 
+  cloudAgent = buildGoModule {
+    pname = "cloud-agent";
+    version = "0.1.0";
+
+    src = fetchFromGitHub {
+      owner = "cdr";
+      repo = "cloud-agent";
+      rev = "0.1.0";
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server/default.nix:33]:
```suggestion
      rev = version;
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/servers/code-server/default.nix (hunk 2 above)
Comment (dguenther) [reply pkgs/servers/code-server/default.nix]: Updated as well, thanks!

**Comment by dguenther** [comment]
This should build now, could I get a re-review?

**Comment by tscholak** [comment]
hey, I’d like to let you know that I was able to adapt the code of this PR for code-server 3.7.2, see https://github.com/hasktorch/hasktorch/blob/4dd5f3b7556e85a2aab47c55f8563e2cbfc70a35/nix/code-server.nix

**Comment by dguenther** [comment]
Oh great, thanks for letting me know! Did you open a PR to nixpkgs with that? Otherwise I'll close this one and open a new one with your changes, if that's okay with you.

**Comment by tscholak** [comment]
no, I haven't opened a new pr for that. I had to modify one `sed` call because it was otherwise destroying the `postinstall.sh` file, and I've tried to make it work on macOS, but so far without success.

**Comment by dguenther** [comment]
I was debating the value of replacing the build steps with a download of the release artifacts, [similar to what the vscode expression does](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vscode/vscode.nix), as it would certainly simplify the whole thing and make macOS compatibility much easier as well. I'm not sure what the general nixpkgs stance is on those types of expressions, since I suppose it would lose the benefit of tracking build dependencies through Nix.

**Comment by tscholak** [comment]
I suppose it would simplify things, but so far this particular from-source derivation is working well. It would definitely be a regression to switch to release artifacts. Besides, I attribute my failure wrt macOS to my own ignorance, I know virtually nothing about building npm code.

**Comment by tscholak** [comment]
btw, here's the error I'm currently stuck with:
```
building
yarn install v1.22.10
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
error Can't make a request in offline mode ("https://registry.yarnpkg.com/bytes/-/bytes-3.1.0.tgz")
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
builder for '/nix/store/3g7khm0qwzwki0jzz6l98rln73idjvwl-code-server-3.7.2.drv' failed with exit code 1
error: build of '/nix/store/3g7khm0qwzwki0jzz6l98rln73idjvwl-code-server-3.7.2.drv' failed
```

**Comment by dguenther** [comment]
Hm, there's probably a macOS-specific yarn package included. Maybe running `nix-build -A code-server.prefetchYarnCache` will return a different `outputHash` for `yarnCache` to use?

**Comment by tscholak** [comment]
yes, that worked, thanks. I got the same hash as for linux, `0860h4x4yvi56452bc67avpiv29l7dkn4579pqki3fybi0fpacqf`. Will try to build the code again on macOS.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 101628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>code-server</li>
  </ul>
</details>

`patchPhase` should not be overridden, use `postPatch` instead.

**Comment by dguenther** [comment]
Pushed an update to use `postPatch` instead.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review. If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>code-server</li>
  </ul>
</details>

did a nix-collect-garbage before and the disk has over 100GB free.
```
  [2/4] Fetching packages...                                                                                                                                                                  error Could not write file "/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/lib/vscode/build/yarn-error.log": "ENOSPC: no space left on device, open '/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/lib/vscode/build/yarn-error.log'"                                                                                                     error An unexpected error occurred: "ENOSPC: no space left on device, mkdir '/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/.cache/yarn/v6/npm-node-abort-controller-1.0.4-4095e41d58b2fae169d2f9892904d603e11c7a39-integrity'".                                                                                                                                 info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.                                                                                                    yarn install v1.22.10                                                                                                                                                                       [1/4] Resolving packages...                                                                                                                                                                 [2/4] Fetching packages...                                                                                                                                                                  error Could not write file "/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/lib/vscode/build/lib/watch/yarn-error.log": "ENOSPC: no space left on device, open '/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/lib/vscode/build/lib/watch/yarn-error.log'"                                                                                 error An unexpected error occurred: "ENOSPC: no space left on device, mkdir '/tmp/nix-build-code-server-3.6.0-x86_64-linux-yarn-cache.drv-0/source/.cache/yarn/v6/npm-vscode-gulp-watch-5.0.2-0060ba8d091284a6fbd7e608aa318a9c1d73b840-integrity'".                                                                                                                                     info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.                                                                                                  cannot build derivation '/nix/store/2qs60rs4rirkb3sv770h54xl289wb34c-code-server-3.6.0.drv': 1 dependencies couldn't be built                                                               cannot build derivation '/nix/store/znzi3rg1z5qfcq6xr9w5hdrw6bw4q95w-env.drv': 1 dependencies couldn't be built 
  ```

**Comment by dguenther** [comment]
I ran `nixpkgs-review pr 101628` locally and it completed successfully for me, would you mind trying it again or helping me out with some suggestions for reproduction steps?

**Comment by contrun** [comment]
Can you please kindly update code server to 3.8.0?

**Comment by dguenther** [comment]
Sure, this may take some time as 3.8.0 brings some changes to the VS Code patch process.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>code-server</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 101628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>code-server</li>
</ul>
</details>

**Pull request merged**